From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: okx-code <okx@okx.sh>
Date: Thu, 6 Feb 2025 19:05:00 +0100
Subject: [PATCH] Paper PR: Optimise temptation lookups

Original license: GPL-3.0-only
Original project: https://github.com/PaperMC/Paper
Paper pull request: https://github.com/PaperMC/Paper/pull/11942

Both TemptGoal and TemptingSensor recheck their validity each tick.
For this, they iterate all online players, checking their main and
offhand items against item tags.
This logic quickly explodes as each brain/goal selector queries all
players.

This patch attempts to optimise this behaviour by lazily caching the
results of the non-entity specific checks in a single BitSet.
Such cache is valid for a single tick but can be re-used by each tempt
goal or sensor sharing the same non-entity specific selection
predicates.

diff --git a/io/papermc/paper/entity/temptation/GlobalTemptationLookup.java b/io/papermc/paper/entity/temptation/GlobalTemptationLookup.java
new file mode 100644
index 0000000000000000000000000000000000000000..c3339b22929cb4e3b216aadf1069daa1f792d74e
--- /dev/null
+++ b/io/papermc/paper/entity/temptation/GlobalTemptationLookup.java
@@ -0,0 +1,87 @@
+package io.papermc.paper.entity.temptation;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.function.Predicate;
+import net.minecraft.tags.ItemTags;
+import net.minecraft.world.entity.animal.frog.FrogAi;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+
+/**
+ * The tempt state lookup holds onto cached temptation flags of players in the world.
+ */
+public class GlobalTemptationLookup {
+    private static int registeredPredicateCounter = 0;
+
+    public static final TemptationPredicate BEE_FOOD = register(i -> i.is(ItemTags.BEE_FOOD));
+    public static final TemptationPredicate CHICKEN_FOOD = register(i -> i.is(ItemTags.CHICKEN_FOOD));
+    public static final TemptationPredicate COW_FOOD = register(i -> i.is(ItemTags.COW_FOOD));
+    public static final TemptationPredicate PANDA_FOOD = register(i -> i.is(ItemTags.PANDA_FOOD));
+    public static final TemptationPredicate PIG_CARROT_ON_A_STICK = register(i -> i.is(Items.CARROT_ON_A_STICK));
+    public static final TemptationPredicate PIG = register(i -> i.is(ItemTags.PIG_FOOD));
+    public static final TemptationPredicate RABBIT_FOOD = register(i -> i.is(ItemTags.RABBIT_FOOD));
+    public static final TemptationPredicate SHEEP_FOOD = register(i -> i.is(ItemTags.SHEEP_FOOD));
+    public static final TemptationPredicate TURTLE_FOOD = register(i -> i.is(ItemTags.TURTLE_FOOD));
+    public static final TemptationPredicate HORSE_FOOD = register(i -> i.is(ItemTags.HORSE_TEMPT_ITEMS));
+    public static final TemptationPredicate LLAMA_TEMPT_ITEMS = register(i -> i.is(ItemTags.LLAMA_TEMPT_ITEMS));
+    public static final TemptationPredicate STRIDER_TEMPT_ITEMS = register(i -> i.is(ItemTags.STRIDER_TEMPT_ITEMS));
+    public static final TemptationPredicate CAT_FOOD = register(i -> i.is(ItemTags.CAT_FOOD));
+    public static final TemptationPredicate OCELOT_FOOD = register(i -> i.is(ItemTags.OCELOT_FOOD));
+    public static final TemptationPredicate FROG_TEMPTATIONS = register(FrogAi.getTemptations());
+    public static final TemptationPredicate CAMEL_TEMPTATIONS = register(i -> i.is(ItemTags.CAMEL_FOOD));
+    public static final TemptationPredicate AXOLOTL_TEMPTATIONS = register(i -> i.is(net.minecraft.tags.ItemTags.AXOLOTL_FOOD));
+    public static final TemptationPredicate GOAT_TEMPTATIONS = register(i -> i.is(net.minecraft.tags.ItemTags.GOAT_FOOD));
+    public static final TemptationPredicate ARMADILLO_TEMPTATIONS = register(i -> i.is(ItemTags.ARMADILLO_FOOD));
+    public static final TemptationPredicate SNIFFER_TEMPTATIONS = register(i -> i.is(ItemTags.SNIFFER_FOOD));
+    public static final TemptationPredicate NAUTILUS_TEMPTATIONS = register(i -> i.is(ItemTags.NAUTILUS_FOOD));
+    // NB: Happy ghast sensors are only used for babies. Adults have dynamic food (harness/snowball) so will not be covered here.
+    public static final TemptationPredicate HAPPY_GHAST_TEMPTATIONS = register(i -> i.is(net.minecraft.tags.ItemTags.HAPPY_GHAST_FOOD));
+
+    public record TemptationPredicate(int index, Predicate<ItemStack> predicate) implements Predicate<ItemStack> {
+
+        @Override
+        public boolean test(final ItemStack itemStack) {
+            return this.predicate.test(itemStack);
+        }
+    }
+
+    public static int indexFor(final Predicate<ItemStack> predicate) {
+        return predicate instanceof final TemptationPredicate temptationPredicate ? temptationPredicate.index() : -1;
+    }
+
+    private static TemptationPredicate register(final Predicate<ItemStack> predicate) {
+        final TemptationPredicate val = new TemptationPredicate(registeredPredicateCounter, predicate);
+        registeredPredicateCounter++;
+        return val;
+    }
+
+    private final List<BitSet> precalculatedTemptItems = new ArrayList<>();
+    private final BitSet calculatedThisTick = new BitSet();
+
+    {
+        for (int i = 0; i < registeredPredicateCounter; i++) {
+            this.precalculatedTemptItems.add(new BitSet());
+        }
+    }
+
+    public void reset() {
+        for (int i = 0; i < registeredPredicateCounter; i++) {
+            this.precalculatedTemptItems.get(i).clear();
+        }
+        this.calculatedThisTick.clear();
+    }
+
+    public boolean isCalculated(final int index) {
+        return this.calculatedThisTick.get(index);
+    }
+
+    public void setCalculated(final int index) {
+        this.calculatedThisTick.set(index);
+    }
+
+    public BitSet getBitSet(final int index) {
+        return this.precalculatedTemptItems.get(index);
+    }
+}
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 558ce0a558b467acdbf3d76e4126142302655cc5..8b64198c633e31204eee7aeb6d843d2ed2551e0e 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1072,6 +1072,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
             boolean didDespawn = tickRateManager.runsNormally() && despawnMap.tick(this, this.entityTickList); // Leaf - optimize despawn
+            this.globalTemptationLookup.reset(); // Paper - optimise temptation lookups - reset global cache prior to next entity tick
             this.entityTickList
                 .forEach(
                     entity -> {
@@ -3545,4 +3546,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return (System.nanoTime() - MinecraftServer.SERVER_INIT) / (java.util.concurrent.TimeUnit.MILLISECONDS.toNanos(50L));
     }
     // Paper end - lag compensation
+
+    // Paper start - optimise temptation lookups
+    private final io.papermc.paper.entity.temptation.GlobalTemptationLookup globalTemptationLookup = new io.papermc.paper.entity.temptation.GlobalTemptationLookup(); // Paper - Optimise TemptGoal
+    public io.papermc.paper.entity.temptation.GlobalTemptationLookup getTemptGoalLookup() {
+        return globalTemptationLookup;
+    }
+    // Paper end - optimise temptation lookups
 }
diff --git a/net/minecraft/world/entity/ai/goal/TemptGoal.java b/net/minecraft/world/entity/ai/goal/TemptGoal.java
index 016167a6ff90dfe5a2e66596eaab2c8f2775c016..d31c0139723f7b9b3fa04f373debc34c722abfeb 100644
--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -29,6 +29,7 @@ public class TemptGoal extends Goal {
     private final Predicate<ItemStack> items;
     private final boolean canScare;
     private final double stopDistance;
+    private final int globalTemptationLookupIndex; // Paper - optimise temptation checks
 
     public TemptGoal(final PathfinderMob mob, final double speedModifier, final Predicate<ItemStack> items, final boolean canScare) {
         this((Mob)mob, speedModifier, items, canScare, 2.5);
@@ -39,13 +40,14 @@ public class TemptGoal extends Goal {
     }
 
     private TemptGoal(final Mob mob, final double speedModifier, final Predicate<ItemStack> items, final boolean canScare, final double stopDistance) {
+        this.globalTemptationLookupIndex = io.papermc.paper.entity.temptation.GlobalTemptationLookup.indexFor(items); // Paper - optimise temptation checks
         this.mob = mob;
         this.speedModifier = speedModifier;
         this.items = items;
         this.canScare = canScare;
         this.stopDistance = stopDistance;
         this.setFlags(EnumSet.of(Goal.Flag.MOVE, Goal.Flag.LOOK));
-        this.targetingConditions = TEMPT_TARGETING.copy().selector((target, level) -> this.shouldFollow(target));
+        this.targetingConditions = globalTemptationLookupIndex >= 0 ? TEMPT_TARGETING.copy() : TEMPT_TARGETING.copy().selector((target, level) -> this.shouldFollow(target)); // Paper - optimise temptation checks - skip selector if we have a lookup index.
     }
 
     @Override
@@ -54,8 +56,41 @@ public class TemptGoal extends Goal {
             this.calmDown--;
             return false;
         } else {
+            // Paper start - optimise temptation lookups
+            final TargetingConditions rangeTargetingConditions = this.targetingConditions.range(this.mob.getAttributeValue(Attributes.TEMPT_RANGE));
+
+            if (this.globalTemptationLookupIndex != -1) {
+                final net.minecraft.server.level.ServerLevel level = getServerLevel(this.mob);
+                final io.papermc.paper.entity.temptation.GlobalTemptationLookup lookup = level.getTemptGoalLookup();
+                final java.util.BitSet lookupBitSet = lookup.getBitSet(this.globalTemptationLookupIndex);
+                final java.util.List<net.minecraft.server.level.ServerPlayer> players = level.players();
+                // Check if the lookup needs to be computed this tick. Do so for all players if needed.
+                if (!lookup.isCalculated(this.globalTemptationLookupIndex)) {
+                    for (int i = 0; i < players.size(); i++) {
+                        lookupBitSet.set(i, shouldFollow(players.get(i)));
+                    }
+                    lookup.setCalculated(this.globalTemptationLookupIndex);
+                }
+                double d = -1.0;
+                net.minecraft.server.level.ServerPlayer nearestPlayer = null;
+                // Only iterate over players that passed #shouldFollow either in the prior computation or another goals canUse check.
+                for (int i = lookupBitSet.nextSetBit(0); i >= 0; i = lookupBitSet.nextSetBit(i + 1)) {
+                    final net.minecraft.server.level.ServerPlayer player = players.get(i);
+                    if (rangeTargetingConditions.test(level, this.mob, player)) {
+                        final double d1 = player.distanceToSqr(this.mob.getX(), this.mob.getY(), this.mob.getZ());
+                        if (d == -1.0 || d1 < d) {
+                            d = d1;
+                            nearestPlayer = player;
+                        }
+                    }
+                }
+                this.player = nearestPlayer;
+            } else {
+            // Default case for non-optimized / non vanilla tempt goal predicates.
+            // Paper end - optimise temptation lookups
             this.player = getServerLevel(this.mob)
                 .getNearestPlayer(this.targetingConditions.range(this.mob.getAttributeValue(Attributes.TEMPT_RANGE)), this.mob);
+            } // Paper - optimise temptation lookups
             // CraftBukkit start
             if (this.player != null) {
                 org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(this.mob, this.player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TEMPT);
diff --git a/net/minecraft/world/entity/ai/sensing/SensorType.java b/net/minecraft/world/entity/ai/sensing/SensorType.java
index 834916d683c03fda057218ec2f562f41139f2b02..393f068b40e02a6f3cb2e3828ddae9b7d1f6e48f 100644
--- a/net/minecraft/world/entity/ai/sensing/SensorType.java
+++ b/net/minecraft/world/entity/ai/sensing/SensorType.java
@@ -35,9 +35,15 @@ public class SensorType<U extends Sensor<?>> {
     public static final SensorType<AdultSensor> NEAREST_ADULT_ANY_TYPE = register("nearest_adult_any_type", AdultSensorAnyType::new);
     public static final SensorType<AxolotlAttackablesSensor> AXOLOTL_ATTACKABLES = register("axolotl_attackables", AxolotlAttackablesSensor::new);
     public static final SensorType<TemptingSensor> FOOD_TEMPTATIONS = register("food_temptations", TemptingSensor::forAnimal);
-    public static final SensorType<TemptingSensor> FROG_TEMPTATIONS = register("frog_temptations", () -> new TemptingSensor(FrogAi.getTemptations()));
+    public static final SensorType<TemptingSensor> AXOLOTL_TEMPTATIONS = register("axolotl_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.AXOLOTL_TEMPTATIONS)); // Paper - optimise temptation lookups
+    public static final SensorType<TemptingSensor> GOAT_TEMPTATIONS = register("goat_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.GOAT_TEMPTATIONS)); // Paper - optimise temptation lookups
+    public static final SensorType<TemptingSensor> CAMEL_TEMPTATIONS = register("camel_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.CAMEL_TEMPTATIONS)); // Paper - optimise temptation lookups
+    public static final SensorType<TemptingSensor> ARMADILLO_TEMPTATIONS = register("armadillo_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.ARMADILLO_TEMPTATIONS)); // Paper - optimise temptation lookups
+    public static final SensorType<TemptingSensor> SNIFFER_TEMPTATIONS = register("sniffer_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.SNIFFER_TEMPTATIONS)); // Paper - optimise temptation lookups
+    public static final SensorType<TemptingSensor> HAPPY_GHAST_TEMPTATIONS = register("happy_ghast_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.HAPPY_GHAST_TEMPTATIONS)); // Paper - optimise temptation lookups
+    public static final SensorType<TemptingSensor> FROG_TEMPTATIONS = register("frog_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.FROG_TEMPTATIONS));
     public static final SensorType<TemptingSensor> NAUTILUS_TEMPTATIONS = register(
-        "nautilus_temptations", () -> new TemptingSensor(NautilusAi.getTemptations())
+        "nautilus_temptations", () -> new TemptingSensor(io.papermc.paper.entity.temptation.GlobalTemptationLookup.NAUTILUS_TEMPTATIONS)
     );
     public static final SensorType<FrogAttackablesSensor> FROG_ATTACKABLES = register("frog_attackables", FrogAttackablesSensor::new);
     public static final SensorType<IsInWaterSensor> IS_IN_WATER = register("is_in_water", IsInWaterSensor::new);
diff --git a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
index 70e5ffafdf8e78efb36c4480c0abe515dc35c62f..b69b35999ed4ed7b5b1d00e1b0d19dcc6862c9bd 100644
--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -21,9 +21,11 @@ import net.minecraft.world.item.ItemStack;
 public class TemptingSensor extends Sensor<PathfinderMob> {
     private static final TargetingConditions TEMPT_TARGETING = TargetingConditions.forNonCombat().ignoreLineOfSight();
     private final BiPredicate<PathfinderMob, ItemStack> temptations;
+    private final int globalTemptationLookupIndex; // Paper - optimise temptation lookups
 
     public TemptingSensor(final Predicate<ItemStack> tt) {
-        this((m, i) -> tt.test(i));
+        this.temptations = (m, i) -> tt.test(i);
+        this.globalTemptationLookupIndex = io.papermc.paper.entity.temptation.GlobalTemptationLookup.indexFor(tt); // Paper - optimise temptation lookups
     }
 
     public static TemptingSensor forAnimal() {
@@ -32,12 +34,45 @@ public class TemptingSensor extends Sensor<PathfinderMob> {
 
     private TemptingSensor(final BiPredicate<PathfinderMob, ItemStack> temptations) {
         this.temptations = temptations;
+        this.globalTemptationLookupIndex = -1;
     }
 
     @Override
     protected void doTick(final ServerLevel level, final PathfinderMob body) {
         Brain<?> brain = body.getBrain();
         TargetingConditions targeting = TEMPT_TARGETING.copy().range((float)body.getAttributeValue(Attributes.TEMPT_RANGE));
+        // Paper start - optimise temptation lookups - on update, ensure below diff filters correctly
+        Player targetPlayer;
+        if (this.globalTemptationLookupIndex != -1) {
+            final io.papermc.paper.entity.temptation.GlobalTemptationLookup lookup = level.getTemptGoalLookup();
+            final java.util.BitSet lookupBitSet = lookup.getBitSet(this.globalTemptationLookupIndex);
+            final java.util.List<net.minecraft.server.level.ServerPlayer> players = level.players();
+            // Check if the lookup needs to be computed this tick. Do so for all players if needed.
+            if (!lookup.isCalculated(this.globalTemptationLookupIndex)) {
+                for (int i = 0; i < players.size(); i++) {
+                    final net.minecraft.server.level.ServerPlayer serverPlayer = players.get(i);
+                    lookupBitSet.set(i, net.minecraft.world.entity.EntitySelector.NO_SPECTATORS.test(serverPlayer) && this.playerHoldingTemptation(body, serverPlayer)); // check on update
+                }
+                lookup.setCalculated(this.globalTemptationLookupIndex);
+            }
+            double d = -1.0;
+            net.minecraft.server.level.ServerPlayer nearestPlayer = null;
+            // Only iterate over players that passed #shouldFollow either in the prior computation or another goals canUse check.
+            for (int i = lookupBitSet.nextSetBit(0); i >= 0; i = lookupBitSet.nextSetBit(i + 1)) {
+                final net.minecraft.server.level.ServerPlayer player = players.get(i);
+                if (targeting.test(level, body, player) && !body.hasPassenger(player)) { // check on update - consider non passengers
+                    final double d1 = player.distanceToSqr(body.getX(), body.getY(), body.getZ());
+                    if (d == -1.0 || d1 < d) {
+                        d = d1;
+                        nearestPlayer = player;
+                    }
+                }
+            }
+            targetPlayer = nearestPlayer;
+        } else {
+            // Default case for non-optimized / non-vanilla tempt goal predicates. Sorting the entire list is completely useless, but none of the vanilla logic uses this path now so
+            // less diff and easier for updates.
+            // Paper end - optimise temptation lookups
         // Leaf start - Remove stream in TemptingSensor
         List<net.minecraft.server.level.ServerPlayer> allPlayers = level.players();
         List<Player> players = new java.util.ArrayList<>();
@@ -52,7 +87,13 @@ public class TemptingSensor extends Sensor<PathfinderMob> {
         // Leaf end - Remove stream in TemptingSensor
         if (!players.isEmpty()) {
             players.sort(Comparator.comparingDouble(body::distanceToSqr)); // Leaf - Remove stream in TemptingSensor
-            Player player = players.get(0);
+            // Paper start - optimise temptation lookups
+        }
+            targetPlayer = players.isEmpty() ? null : players.getFirst();
+        }
+        if (targetPlayer != null) {
+            Player player = targetPlayer;
+            // Paper end - optimise temptation lookups
             // CraftBukkit start
             org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(
                 body, player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TEMPT
diff --git a/net/minecraft/world/entity/animal/armadillo/Armadillo.java b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
index 9f5c072903651aa5f869525157db93aeb63557a2..9f0d4e12b258f01282a5f1a8e12d0b457a376263 100644
--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -64,7 +64,7 @@ public class Armadillo extends Animal {
     private static final EntityDimensions BABY_DIMENSIONS = EntityTypes.ARMADILLO.getDimensions().scale(0.6F).withEyeHeight(0.21875F);
     private static final Brain.Provider<Armadillo> BRAIN_PROVIDER = Brain.provider(
         List.of(
-            SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.FOOD_TEMPTATIONS, SensorType.NEAREST_ADULT, SensorType.ARMADILLO_SCARE_DETECTED
+            SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.ARMADILLO_TEMPTATIONS, SensorType.NEAREST_ADULT, SensorType.ARMADILLO_SCARE_DETECTED
         ),
         var0 -> ArmadilloAi.getActivities()
     );
diff --git a/net/minecraft/world/entity/animal/axolotl/Axolotl.java b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
index 2a0f4648d50581cc9fd58a772cb672eb99f26765..265d6c58e38e1c520d9ef63c2d4e2b8f9f556e7b 100644
--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -77,7 +77,7 @@ public class Axolotl extends Animal implements Bucketable {
     public static final int TOTAL_PLAYDEAD_TIME = 200;
     private static final int POSE_ANIMATION_TICKS = 10;
     private static final Brain.Provider<Axolotl> BRAIN_PROVIDER = Brain.provider(
-        List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_ADULT, SensorType.HURT_BY, SensorType.AXOLOTL_ATTACKABLES, SensorType.FOOD_TEMPTATIONS),
+        List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_ADULT, SensorType.HURT_BY, SensorType.AXOLOTL_ATTACKABLES, SensorType.AXOLOTL_TEMPTATIONS),
         var0 -> AxolotlAi.getActivities()
     );
     private static final EntityDataAccessor<Integer> DATA_VARIANT = SynchedEntityData.defineId(Axolotl.class, EntityDataSerializers.INT);
diff --git a/net/minecraft/world/entity/animal/bee/Bee.java b/net/minecraft/world/entity/animal/bee/Bee.java
index 65a3cfb469fe3a33316aac41680ba216a57a9a3c..4c52d5450d7abc38e3f716652aac23207e34d0b1 100644
--- a/net/minecraft/world/entity/animal/bee/Bee.java
+++ b/net/minecraft/world/entity/animal/bee/Bee.java
@@ -259,7 +259,7 @@ public class Bee extends Animal implements NeutralMob {
         this.goalSelector.addGoal(0, new Bee.BeeAttackGoal(this, 1.4F, true));
         this.goalSelector.addGoal(1, new Bee.BeeEnterHiveGoal());
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, i -> i.is(ItemTags.BEE_FOOD), false));
+        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, io.papermc.paper.entity.temptation.GlobalTemptationLookup.BEE_FOOD, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(3, new Bee.ValidateHiveGoal());
         this.goalSelector.addGoal(3, new Bee.ValidateFlowerGoal());
         this.beePollinateGoal = new Bee.BeePollinateGoal();
diff --git a/net/minecraft/world/entity/animal/camel/Camel.java b/net/minecraft/world/entity/animal/camel/Camel.java
index d6ee82e52487912b131265014d44d78e2625da5b..6d67efaa12d8a1440b7ba265a730c2e0fc4886eb 100644
--- a/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/net/minecraft/world/entity/animal/camel/Camel.java
@@ -73,7 +73,7 @@ public class Camel extends AbstractHorse {
     private static final float SITTING_HEIGHT_DIFFERENCE = 1.43F;
     private static final long DEFAULT_LAST_POSE_CHANGE_TICK = 0L;
     private static final Brain.Provider<Camel> BRAIN_PROVIDER = Brain.provider(
-        List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.FOOD_TEMPTATIONS, SensorType.NEAREST_ADULT), var0 -> CamelAi.getActivities()
+        List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.CAMEL_TEMPTATIONS, SensorType.NEAREST_ADULT), var0 -> CamelAi.getActivities()
     );
     public static final EntityDataAccessor<Boolean> DASH = SynchedEntityData.defineId(Camel.class, EntityDataSerializers.BOOLEAN);
     public static final EntityDataAccessor<Long> LAST_POSE_CHANGE_TICK = SynchedEntityData.defineId(Camel.class, EntityDataSerializers.LONG);
diff --git a/net/minecraft/world/entity/animal/chicken/Chicken.java b/net/minecraft/world/entity/animal/chicken/Chicken.java
index 051be366707f34d4ec58e42e8717c97f308c0ac9..6a82f35b0e3de83f4448e0e22a903291050f048d 100644
--- a/net/minecraft/world/entity/animal/chicken/Chicken.java
+++ b/net/minecraft/world/entity/animal/chicken/Chicken.java
@@ -155,7 +155,7 @@ public class Chicken extends Animal {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         //this.goalSelector.addGoal(1, new PanicGoal(this, 1.4)); // Purpur - Chickens can retaliate - moved down
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(3, new TemptGoal(this, 1.0, i -> i.is(ItemTags.CHICKEN_FOOD), false));
+        this.goalSelector.addGoal(3, new TemptGoal(this, 1.0, io.papermc.paper.entity.temptation.GlobalTemptationLookup.CHICKEN_FOOD, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(4, new FollowParentGoal(this, 1.1));
         this.goalSelector.addGoal(5, new WaterAvoidingRandomStrollGoal(this, 1.0));
         this.goalSelector.addGoal(6, new LookAtPlayerGoal(this, Player.class, 6.0F));
diff --git a/net/minecraft/world/entity/animal/cow/AbstractCow.java b/net/minecraft/world/entity/animal/cow/AbstractCow.java
index f53e44dcb737a8a9fd1c33f158717a11ae399acd..c0bf417e8f84d4995d9a64e25e3f731d267dbb83 100644
--- a/net/minecraft/world/entity/animal/cow/AbstractCow.java
+++ b/net/minecraft/world/entity/animal/cow/AbstractCow.java
@@ -42,7 +42,7 @@ public abstract class AbstractCow extends Animal {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(1, new PanicGoal(this, 2.0));
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, i -> level().purpurConfig.cowFeedMushrooms > 0 && (i.is(net.minecraft.world.level.block.Blocks.RED_MUSHROOM.asItem()) || i.is(net.minecraft.world.level.block.Blocks.BROWN_MUSHROOM.asItem())) || i.is(ItemTags.COW_FOOD), false)); // Purpur - Cows eat mushrooms
+        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, io.papermc.paper.entity.temptation.GlobalTemptationLookup.COW_FOOD, false)); // Purpur - Cows eat mushrooms // Paper - optimise temptation lookups
         this.goalSelector.addGoal(4, new FollowParentGoal(this, 1.25));
         this.goalSelector.addGoal(5, new WaterAvoidingRandomStrollGoal(this, 1.0));
         this.goalSelector.addGoal(6, new LookAtPlayerGoal(this, Player.class, 6.0F));
diff --git a/net/minecraft/world/entity/animal/equine/AbstractHorse.java b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
index 069f970c36a72ff57eb0cf03cf79b1922776061f..beb2b2da1818e7aa9514efe6fc9f927a10c0a497 100644
--- a/net/minecraft/world/entity/animal/equine/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
@@ -200,7 +200,7 @@ public abstract class AbstractHorse extends Animal implements PlayerRideableJump
     protected void addBehaviourGoals() {
         this.goalSelector.addGoal(0, new FloatGoal(this));
         this.goalSelector.addGoal(1, new AbstractHorse.MountPanicGoal(1.2));
-        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, i -> i.is(ItemTags.HORSE_TEMPT_ITEMS), false));
+        this.goalSelector.addGoal(3, new TemptGoal(this, 1.25, io.papermc.paper.entity.temptation.GlobalTemptationLookup.HORSE_FOOD, false)); // Paper - optimise temptation lookups
     }
 
     @Override
diff --git a/net/minecraft/world/entity/animal/equine/Llama.java b/net/minecraft/world/entity/animal/equine/Llama.java
index 0a9dcfcd4c200166b41ab48610f1dc0c7ca90dca..9e6058f5d902a4f11d0a777ece66134a30c764c7 100644
--- a/net/minecraft/world/entity/animal/equine/Llama.java
+++ b/net/minecraft/world/entity/animal/equine/Llama.java
@@ -232,7 +232,7 @@ public class Llama extends AbstractChestedHorse implements RangedAttackMob {
         this.goalSelector.addGoal(3, new RangedAttackGoal(this, 1.25, 40, 20.0F));
         this.goalSelector.addGoal(3, new PanicGoal(this, 1.2));
         this.goalSelector.addGoal(4, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(5, new TemptGoal(this, 1.25, i -> i.is(ItemTags.LLAMA_TEMPT_ITEMS), false));
+        this.goalSelector.addGoal(5, new TemptGoal(this, 1.25, io.papermc.paper.entity.temptation.GlobalTemptationLookup.LLAMA_TEMPT_ITEMS, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(6, new FollowParentGoal(this, 1.0));
         this.goalSelector.addGoal(7, new WaterAvoidingRandomStrollGoal(this, 0.7));
         this.goalSelector.addGoal(8, new LookAtPlayerGoal(this, Player.class, 6.0F));
diff --git a/net/minecraft/world/entity/animal/feline/Cat.java b/net/minecraft/world/entity/animal/feline/Cat.java
index 3414d950bfaf1fbad9c37fb0f5f6fe080ca4294b..e56b902246ee0a81fc40bb52ece9f471fcc03a0c 100644
--- a/net/minecraft/world/entity/animal/feline/Cat.java
+++ b/net/minecraft/world/entity/animal/feline/Cat.java
@@ -175,7 +175,7 @@ public class Cat extends TamableAnimal {
 
     @Override
     protected void registerGoals() {
-        this.temptGoal = new Cat.CatTemptGoal(this, 0.6, i -> i.is(ItemTags.CAT_FOOD), true);
+        this.temptGoal = new Cat.CatTemptGoal(this, 0.6, io.papermc.paper.entity.temptation.GlobalTemptationLookup.CAT_FOOD, true); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(1, new FloatGoal(this));
         this.goalSelector.addGoal(1, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(1, new TamableAnimal.TamableAnimalPanicGoal(1.5));
diff --git a/net/minecraft/world/entity/animal/feline/Ocelot.java b/net/minecraft/world/entity/animal/feline/Ocelot.java
index 6f89312c7746057bc73fde32c95af348b651989f..4182438b905d4eac729436e30dbd4559844e7860 100644
--- a/net/minecraft/world/entity/animal/feline/Ocelot.java
+++ b/net/minecraft/world/entity/animal/feline/Ocelot.java
@@ -165,7 +165,7 @@ public class Ocelot extends Animal {
 
     @Override
     protected void registerGoals() {
-        this.temptGoal = new Ocelot.OcelotTemptGoal(this, 0.6, i -> i.is(ItemTags.OCELOT_FOOD), true);
+        this.temptGoal = new Ocelot.OcelotTemptGoal(this, 0.6, io.papermc.paper.entity.temptation.GlobalTemptationLookup.OCELOT_FOOD, true); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(1, new FloatGoal(this));
         this.goalSelector.addGoal(1, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(3, this.temptGoal);
diff --git a/net/minecraft/world/entity/animal/goat/Goat.java b/net/minecraft/world/entity/animal/goat/Goat.java
index 30d9800bd4a0e4a0a6a2bd0e95151cbf3d6eaea5..1d54c17b1a7f90d5d5edcc151c7bb367cb579ea4 100644
--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -72,7 +72,7 @@ public class Goat extends Animal {
             SensorType.NEAREST_ITEMS,
             SensorType.NEAREST_ADULT,
             SensorType.HURT_BY,
-            SensorType.FOOD_TEMPTATIONS
+            SensorType.GOAT_TEMPTATIONS
         ),
         var0 -> GoatAi.getActivities()
     );
diff --git a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
index 471a6a764646cc40761d320a54ff540d565b35d3..54ad7be059b7981e10c40ad56fb0ecaac6f04d90 100644
--- a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
+++ b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
@@ -66,7 +66,7 @@ public class HappyGhast extends Animal {
     private static final int MAX_STILL_TIMEOUT = 10;
     private static final Brain.Provider<HappyGhast> BRAIN_PROVIDER = Brain.provider(
         List.of(
-            SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.FOOD_TEMPTATIONS, SensorType.NEAREST_ADULT_ANY_TYPE, SensorType.NEAREST_PLAYERS
+            SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.HAPPY_GHAST_TEMPTATIONS, SensorType.NEAREST_ADULT_ANY_TYPE, SensorType.NEAREST_PLAYERS
         ),
         var0 -> HappyGhastAi.getActivities()
     );
diff --git a/net/minecraft/world/entity/animal/panda/Panda.java b/net/minecraft/world/entity/animal/panda/Panda.java
index 4b3e3b3647950e304a4d8ac685760f714548cd36..de85704fe6c2b7b787721ebda99995ed41455305 100644
--- a/net/minecraft/world/entity/animal/panda/Panda.java
+++ b/net/minecraft/world/entity/animal/panda/Panda.java
@@ -341,7 +341,7 @@ public class Panda extends Animal {
         this.goalSelector.addGoal(2, new Panda.PandaPanicGoal(this, 2.0));
         this.goalSelector.addGoal(2, new Panda.PandaBreedGoal(this, 1.0));
         this.goalSelector.addGoal(3, new Panda.PandaAttackGoal(this, 1.2F, true));
-        this.goalSelector.addGoal(4, new TemptGoal(this, 1.0, i -> i.is(ItemTags.PANDA_FOOD), false));
+        this.goalSelector.addGoal(4, new TemptGoal(this, 1.0, io.papermc.paper.entity.temptation.GlobalTemptationLookup.PANDA_FOOD, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(6, new Panda.PandaAvoidGoal<>(this, Player.class, 8.0F, 2.0, 2.0));
         this.goalSelector.addGoal(6, new Panda.PandaAvoidGoal<>(this, Monster.class, 4.0F, 2.0, 2.0));
         this.goalSelector.addGoal(7, new Panda.PandaSitGoal());
diff --git a/net/minecraft/world/entity/animal/pig/Pig.java b/net/minecraft/world/entity/animal/pig/Pig.java
index fdb6b128bbc82ba36717afc48a06f2a9cd1ed78a..cb09000e6c778c0fb995377b6d6c14a94341b690 100644
--- a/net/minecraft/world/entity/animal/pig/Pig.java
+++ b/net/minecraft/world/entity/animal/pig/Pig.java
@@ -145,8 +145,8 @@ public class Pig extends Animal implements ItemSteerable {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(1, new PanicGoal(this, 1.25));
         this.goalSelector.addGoal(3, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(4, new TemptGoal(this, 1.2, i -> i.is(Items.CARROT_ON_A_STICK), false));
-        this.goalSelector.addGoal(4, new TemptGoal(this, 1.2, i -> i.is(ItemTags.PIG_FOOD), false));
+        this.goalSelector.addGoal(4, new TemptGoal(this, 1.2, io.papermc.paper.entity.temptation.GlobalTemptationLookup.PIG_CARROT_ON_A_STICK, false)); // Paper - optimise temptation lookups
+        this.goalSelector.addGoal(4, new TemptGoal(this, 1.2, io.papermc.paper.entity.temptation.GlobalTemptationLookup.PIG, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(5, new FollowParentGoal(this, 1.1));
         this.goalSelector.addGoal(6, new WaterAvoidingRandomStrollGoal(this, 1.0));
         this.goalSelector.addGoal(7, new LookAtPlayerGoal(this, Player.class, 6.0F));
diff --git a/net/minecraft/world/entity/animal/rabbit/Rabbit.java b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
index 6b487cb7c0bfaa882df6c7506350171a2d01d216..db5b6f673e4726ca6d88744d9348ee3679c9db17 100644
--- a/net/minecraft/world/entity/animal/rabbit/Rabbit.java
+++ b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
@@ -216,7 +216,7 @@ public class Rabbit extends Animal {
         this.goalSelector.addGoal(1, new ClimbOnTopOfPowderSnowGoal(this, this.level()));
         this.goalSelector.addGoal(1, new Rabbit.RabbitPanicGoal(this, 2.2));
         this.goalSelector.addGoal(2, new BreedGoal(this, 0.8));
-        this.goalSelector.addGoal(3, new TemptGoal(this, 1.0, i -> i.is(ItemTags.RABBIT_FOOD), false));
+        this.goalSelector.addGoal(3, new TemptGoal(this, 1.0, io.papermc.paper.entity.temptation.GlobalTemptationLookup.RABBIT_FOOD, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(4, new Rabbit.RabbitAvoidEntityGoal<>(this, Player.class, 8.0F, 2.2, 2.2));
         this.goalSelector.addGoal(4, new Rabbit.RabbitAvoidEntityGoal<>(this, Wolf.class, 10.0F, 2.2, 2.2));
         this.goalSelector.addGoal(4, new Rabbit.RabbitAvoidEntityGoal<>(this, Monster.class, 4.0F, 2.2, 2.2));
diff --git a/net/minecraft/world/entity/animal/sheep/Sheep.java b/net/minecraft/world/entity/animal/sheep/Sheep.java
index c062cff71a6d2ecc5bf9492715a6c9947b5129ea..9a2ecdd656fb9166dd94525ec57a7e4369a8a87a 100644
--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -141,7 +141,7 @@ public class Sheep extends Animal implements Shearable {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(1, new PanicGoal(this, 1.25));
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(3, new TemptGoal(this, 1.1, i -> i.is(ItemTags.SHEEP_FOOD), false));
+        this.goalSelector.addGoal(3, new TemptGoal(this, 1.1, io.papermc.paper.entity.temptation.GlobalTemptationLookup.SHEEP_FOOD, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(4, new FollowParentGoal(this, 1.1));
         this.goalSelector.addGoal(5, this.eatBlockGoal);
         this.goalSelector.addGoal(6, new WaterAvoidingRandomStrollGoal(this, 1.0));
diff --git a/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index 1780dac959aabadf6ae8ab05063da83faf99a71e..90884f7c6f4e16bc16b32532c5943032f9b54441 100644
--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -64,7 +64,7 @@ import net.minecraft.world.phys.Vec3;
 public class Sniffer extends Animal {
     private static final Brain.Provider<Sniffer> BRAIN_PROVIDER = Brain.provider(
         List.of(MemoryModuleType.SNIFFER_EXPLORED_POSITIONS),
-        List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.NEAREST_PLAYERS, SensorType.FOOD_TEMPTATIONS),
+        List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.HURT_BY, SensorType.NEAREST_PLAYERS, SensorType.SNIFFER_TEMPTATIONS),
         var0 -> SnifferAi.getActivities()
     );
     private static final int DIGGING_PARTICLES_DELAY_TICKS = 1700;
diff --git a/net/minecraft/world/entity/animal/turtle/Turtle.java b/net/minecraft/world/entity/animal/turtle/Turtle.java
index d1edd97b879333a9fee4e4e9ea3f3f6af39006af..64d268d84d91579efb61b288bc520385a90009a9 100644
--- a/net/minecraft/world/entity/animal/turtle/Turtle.java
+++ b/net/minecraft/world/entity/animal/turtle/Turtle.java
@@ -200,7 +200,7 @@ public class Turtle extends Animal {
         this.goalSelector.addGoal(0, new Turtle.TurtlePanicGoal(this, 1.2));
         this.goalSelector.addGoal(1, new Turtle.TurtleBreedGoal(this, 1.0));
         this.goalSelector.addGoal(1, new Turtle.TurtleLayEggGoal(this, 1.0));
-        this.goalSelector.addGoal(2, new TemptGoal(this, 1.1, i -> i.is(ItemTags.TURTLE_FOOD), false));
+        this.goalSelector.addGoal(2, new TemptGoal(this, 1.1, io.papermc.paper.entity.temptation.GlobalTemptationLookup.TURTLE_FOOD, false)); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(3, new Turtle.TurtleGoToWaterGoal(this, 1.0));
         this.goalSelector.addGoal(4, new Turtle.TurtleGoHomeGoal(this, 1.0));
         this.goalSelector.addGoal(7, new Turtle.TurtleTravelGoal(this, 1.0));
diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index d612b2fd802ba951ecfdfcf71b1c7cdf61fd9e66..22081ffb051315bacbf52454303df39ecfa355ef 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -207,7 +207,7 @@ public class Strider extends Animal implements ItemSteerable {
         this.goalSelector.addGoal(1, new PanicGoal(this, 1.65));
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
-        this.temptGoal = new TemptGoal(this, 1.4, i -> i.is(ItemTags.STRIDER_TEMPT_ITEMS), false);
+        this.temptGoal = new TemptGoal(this, 1.4, io.papermc.paper.entity.temptation.GlobalTemptationLookup.STRIDER_TEMPT_ITEMS, false); // Paper - optimise temptation lookups
         this.goalSelector.addGoal(3, this.temptGoal);
         this.goalSelector.addGoal(4, new Strider.StriderGoToLavaGoal(this, 1.0));
         this.goalSelector.addGoal(5, new FollowParentGoal(this, 1.0));
